### PR TITLE
[cluster] Close etcd watches completely before recreating

### PR DESCRIPTION
**Pull request by vdarulis**
**What this PR does / why we need it**:

This PR fixes occasional complete watch hangs if etcd cluster loses the leader node and/or is unavailable for some period of time.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/cluster/etcd/watchmanager/manager.go [hunk 1]
```diff
@@ -71,27 +71,39 @@ func (w *manager) watchChanWithTimeout(key string, rev int64) (clientv3.WatchCha
...
+	var (
+		watcher   = clientv3.NewWatcher(w.opts.Client())
+		watchChan clientv3.WatchChan
+	)
 	go func() {
 		wOpts := w.opts.WatchOptions()
 		if rev > 0 {
 			wOpts = append(wOpts, clientv3.WithRev(rev))
 		}
-		watchChan = w.opts.Watcher().Watch(
```
Comment (abliqo) [inline src/cluster/etcd/watchmanager/manager.go:80]: nit: should we add comment to the definition of `Watcher()` to not rely on it?

File: src/cluster/etcd/watchmanager/manager.go [hunk 2]
```diff
@@ -71,27 +71,39 @@ func (w *manager) watchChanWithTimeout(key string, rev int64) (clientv3.WatchCha
 
 	ctx, cancelFn := context.WithCancel(clientv3.WithRequireLeader(context.Background()))
 
-	var watchChan clientv3.WatchChan
+	var (
+		watcher   = clientv3.NewWatcher(w.opts.Client())
```
Comment (abliqo) [inline src/cluster/etcd/watchmanager/manager.go:75]: can't find the definition fo `Client()` in `w.opts`

File: src/cluster/etcd/watchmanager/manager.go [hunk 3]
```diff
@@ -108,6 +120,16 @@ func (w *manager) Watch(key string) {
 
 	defer ticker.Stop()
 
+	resetWatch := func() {
```
Comment (abliqo) [inline src/cluster/etcd/watchmanager/manager.go:123]: ultra-nit: perhaps naming it `resetWatchAndSleep` would be more accurate.

File: src/cluster/etcd/watchmanager/manager.go (hunk 2 above)
Comment (abliqo) [reply src/cluster/etcd/watchmanager/manager.go:75]: my bad, didn't see it below

**Review by abliqo — APPROVED** [review]

**Review by abliqo — APPROVED** [review]

File: src/cluster/etcd/watchmanager/manager.go [hunk 4]
```diff
@@ -71,27 +71,39 @@ func (w *manager) watchChanWithTimeout(key string, rev int64) (clientv3.WatchCha
...
 		close(doneCh)
 	}()
 
-	timeout := w.opts.WatchChanInitTimeout()
+	var (
+		timeout       = w.opts.WatchChanInitTimeout()
+		cancelWatchFn = func() {
+			cancelFn()
+			if err := watcher.Close(); err != nil {
+				w.logger.Info("error closing watcher", zap.Error(err))
```
Comment (cw9) [inline src/cluster/etcd/watchmanager/manager.go:96]: I'd add a comment that why this error is ok to ignore

File: src/cluster/etcd/watchmanager/manager.go [hunk 5]
```diff
@@ -71,27 +71,39 @@ func (w *manager) watchChanWithTimeout(key string, rev int64) (clientv3.WatchCha
...
+	)
+
 	select {
 	case <-doneCh:
-		return watchChan, cancelFn, nil
+		return watchChan, cancelWatchFn, nil
 	case <-time.After(timeout):
-		cancelFn()
-		return nil, nil, fmt.Errorf("etcd watch create timed out after %s for key: %s", timeout.String(), key)
+		err := fmt.Errorf("etcd watch create timed out after %s for key: %s", timeout.String(), key)
```
Comment (cw9) [inline src/cluster/etcd/watchmanager/manager.go:105]: can you call cancelWatchFn() here instead of having user to call this on err != nil?

**Pull request merged**